### PR TITLE
feat(replay): lock accounts and resolve lookup table addresses

### PR DESCRIPTION
**Pull request by dnut**

### CODE REVIEW

File: src/replay/account_locks.zig [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        const index = locks.getIndex(address) orelse {
+            return 1;
+        };
+
+        const value = &locks.entries.slice().items(.value)[index];
+        if (value.* == 0) {
+            // this means there is an internal bug within the unlock
+            // method, since the next block here should remove the item
+            // before this number would ever reach zero.
+            unreachable;
```
[the remark this answers is not in the document]
Comment (dnut) [reply src/replay/account_locks.zig:135]: The code in AccountLocks guarantees that it is impossible to reach this line. I think this is a standard usage of `unreachable`.

I mean yeah, if you circumvent the methods and directly mutate the struct's fields to corrupt its state, that's a workaround to reach this line of code. But `unreachable` is still commonly used to verify that state is not corrupted. The contract is, "if you misuse this struct, it has undefined behavior."

----

This code is equivalent to writing `assert(value.* != 0)`. It's a safety check that's just here to verify an assumed invariant. Using `unreachable` means that the check is only run in build modes where safety checks are enabled. Switching to a panic means the check will run even in build modes where safety checks are disabled.

I lean towards allowing this to be optimized away if someone is really willing to disable safety checks. Personally I never compile sig without safety checks.

We could write our own assert function that uses panic instead of unreachable, and replace existing usages of `std.debug.assert` with our own assert, to force assertions to run even when safety checks are disabled.

File: src/replay/resolve_lookup.zig [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+pub const ResolvedTransaction = struct {
+    transaction: Transaction,
+    accounts: std.MultiArrayList(InstructionAccount),
+    instructions: []const InstructionInfo,
+
+    pub fn deinit(self: ResolvedTransaction, allocator: Allocator) void {
+        var acc = self.accounts;
+        acc.deinit(allocator);
+        allocator.free(self.instructions);
+        // transaction is not freed because it is borrowed.
```
[the remark this answers is not in the document]
Comment (dnut) [reply src/replay/resolve_lookup.zig:50]: That's one more lifetime to juggle, and I don't see the benefit.

You would need to make sure that you have some safe pre-existing storage for the Transaction struct. The pointers within the `Transaction` should last at least as long as wherever the `Transaction` itself is stored. You'd likely reduce the lifetime of `ResolvedTransaction` by doing this. This constrains the design space and makes the code more prone to memory errors.

File: src/replay/account_locks.zig (hunk 1 above)
[the remark this answers is not in the document]
Comment (dnut) [reply src/replay/account_locks.zig:135]: I have some dead code in here. I don't need this many methods. I'll clean it up which should help narrow the scope that needs to be observed.

File: src/replay/account_locks.zig [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    pub fn unlock(self: *AccountLocks, accounts: []const LockableAccount) u64 {
+        var already_unlocked: u64 = 0;
+        for (accounts) |account| {
+            const locks = if (account.writable) &self.write_locks else &self.readonly_locks;
+            already_unlocked += unlockOneGeneric(locks, account.address);
+        }
+        return already_unlocked;
+    }
+
+    /// assumes that
```
[the remark this answers is not in the document]
Comment (dnut) [reply src/replay/account_locks.zig:110]: This function shouldn't even be here, I forgot about it

[the remark this answers is not in the document]
Comment (dnut) [reply src/replay/account_locks.zig]: d9061f1cd985d6231af826105fbf5f6b0c7755eb

File: src/replay/resolve_lookup.zig [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+            .map => {
+                const map: *const std.AutoArrayHashMapUnmanaged(Pubkey, AddressLookupTable) =
+                    table_provider;
+                return map.get(table_address.*) orelse return error.PubkeyNotInIndex;
+            },
+        }
+    }
+};
+
+test "resolve lookups" {
```
[the remark this answers is not in the document]
Comment (dnut) [reply src/replay/resolve_lookup.zig]: ba2c88014eed587053ee0dd06801d2c47610cd77

File: src/replay/account_locks.zig [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+            .{ .address = test_keys[1], .writable = false },
+            .{ .address = test_keys[1], .writable = false },
+            .{ .address = test_keys[1], .writable = false },
+            .{ .address = test_keys[2], .writable = false },
+            .{ .address = test_keys[3], .writable = true },
+        }));
+    }
+}
+
+test "lockStrict is strict" {
```
[the remark this answers is not in the document]
Comment (yewman) [reply src/replay/account_locks.zig:199]: I guess from my perspective it seemed like duplication of information since the function name is pretty informative. But I understand what your saying about other test implementations!

File: src/replay/account_locks.zig [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    Pubkey.parseBase58String("DttWaMuVvTiduZRnguLF7jNxTgiMBZ1hyAumKUiL2KRL") catch unreachable,
+    Pubkey.parseBase58String("9fRXX5Bj3XWfCHtVkYtQVvMnAfoy4KjcpgHTBSmymzRu") catch unreachable,
+    Pubkey.parseBase58String("5BUYHtAdv2rUM73A8iEWQ4hcVQXCRE8VrQFR39DKaWW8") catch unreachable,
+    Pubkey.parseBase58String("9jo7RYY8HgxpU3Zs5zPFyRAkcUtx8J5RHc2fN9Btxfmi") catch unreachable,
+    Pubkey.parseBase58String("FZgqDx8Guf649hif1WVuTb6mTV2LhotUFE12PYveAbC8") catch unreachable,
+    Pubkey.parseBase58String("JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4") catch unreachable,
+    Pubkey.parseBase58String("pAMMBay6oceH9fJKBRHGP5D4bD4sWpmSwMn52FMfXEA") catch unreachable,
+};
+
+test "locking/unlocking basically works" {
```
[the remark this answers is not in the document]
Comment (yewman) [reply src/replay/account_locks.zig:146]: yeah thats fair, can we remove `basically works` though? If more testing is necessary I think a TODO would be more appropriate. `basically works` leaves me wondering if we need to test this more thoroughly or if it is ok as is

File: src/replay/resolve_lookup.zig (hunk 2 above)
[the remark this answers is not in the document]
Comment (yewman) [reply src/replay/resolve_lookup.zig:50]: Yeah that is fair, was just curious 😛

File: src/replay/account_locks.zig (hunk 6 above)
[the remark this answers is not in the document]
Comment (dnut) [reply src/replay/account_locks.zig]: I see your point, the name sounds too pessimistic. This test is intended to cover some ordinary happy paths and failure modes without going deep into complexity or edge cases. If I wanted to implement some more specific testing strategies, I would do that in separate tests, like the other ones in this file.

I'm thinking a better name is "lock and unlock: simple happy paths and failure modes"

We can always add more test coverage, but in my opinion, the coverage is satisfactory for now. I don't think a todo is necessary since the same (or worse) can be said for most of the codebase.

[the remark this answers is not in the document]
Comment (dnut) [reply src/replay/account_locks.zig]: 935ceb0829f19515ec2dcec3b593454f82f7d4ea

**Review by yewman — APPROVED** [review]

File: src/replay/account_locks.zig [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+const Allocator = std.mem.Allocator;
+
+const Pubkey = sig.core.Pubkey;
+
+pub const LockableAccount = struct { address: Pubkey, writable: bool };
+
+/// Only interact with this struct using its public methods. Directly mutating
+/// the fields may result in undefined behavior.
+pub const AccountLocks = struct {
+    write_locks: std.AutoArrayHashMapUnmanaged(Pubkey, u64) = .{},
```
Comment (dadepo) [inline src/replay/account_locks.zig:13]: I think `HashMapUnmanaged` might be slightly better as I believe that is optimized for lookup times.

File: src/replay/account_locks.zig [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    readonly_locks: std.AutoArrayHashMapUnmanaged(Pubkey, u64) = .{},
+
+    const LockError = Allocator.Error || error{LockFailed};
+
+    pub fn deinit(self: AccountLocks, allocator: Allocator) void {
+        var write_locks = self.write_locks;
```
Comment (dadepo) [inline src/replay/account_locks.zig:19]: Should the values within the map not also be iterated and `denited`?
